pallets/inflation-rewards/src/lib.rs

```diff
@@ -0,0 +1,316 @@
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+use frame_support::{
+    pallet_prelude::*,
+    traits::{
+        fungible::{Balanced, Credit, Inspect},
+        tokens::{Fortitude, Precision, Preservation},
+        Imbalance, OnUnbalanced,
+    },
+};
+use frame_system::pallet_prelude::*;
+use sp_runtime::{traits::Get, Perbill};
+use tp_core::{BlockNumber, ParaId};
+use tp_traits::{AuthorNotingHook, DistributeRewards, GetCurrentContainerChains};
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+
+    pub type BalanceOf<T> =
+        <<T as Config>::Currency as Inspect<<T as frame_system::Config>::AccountId>>::Balance;
+    pub type CreditOf<T> = Credit<<T as frame_system::Config>::AccountId, <T as Config>::Currency>;
+
+    /// Inflation rewards pallet.
+    #[pallet::pallet]
+    #[pallet::without_storage_info]
+    pub struct Pallet<T>(core::marker::PhantomData<T>);
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+        fn on_initialize(_: T::BlockNumber) -> Weight {
+            let mut weight = T::DbWeight::get().reads(1);
+
+            // Collect indistributed rewards, if any
+            let not_distributed_rewards =
+                if let Some(chains_to_reward) = ChainsToReward::<T>::take() {
+                    // Collect and sum all undistributed rewards
+                    let rewards_not_distributed: BalanceOf<T> = chains_to_reward.rewards_per_chain
+                        * (chains_to_reward.para_ids.len() as u32).into();
+                    T::Currency::withdraw(
+                        &T::PendingRewardsAccount::get(),
+                        rewards_not_distributed,
+                        Precision::BestEffort,
+                        Preservation::Expendable,
+                        Fortitude::Force,
+                    )
+                    .unwrap_or(CreditOf::<T>::zero())
+                } else {
+                    CreditOf::<T>::zero()
+                };
+
+            // Get the number of chains at this block
+            weight += T::DbWeight::get().reads_writes(1, 1);
+            let registered_para_ids = T::ContainerChains::current_container_chains();
+
+            let number_of_chains: BalanceOf<T> =
+                ((registered_para_ids.len() as u32).saturating_add(1)).into();
+
+            // Issue new supply
+            let new_supply =
+                T::Currency::issue(T::InflationRate::get() * T::Currency::total_issuance());
+
+            // Split staking reward portion
+            let total_rewards = T::RewardsPortion::get() * new_supply.peek();
+            let (rewards_credit, reminder_credit) = new_supply.split(total_rewards);
+            let rewards_per_chain: BalanceOf<T> = rewards_credit.peek() / number_of_chains;
+            let (mut total_reminder, staking_rewards) = rewards_credit.split_merge(
```

Comment:
what does split_merge do?

Reply:
In substrate code:
```rust
fn split_merge(self, amount: Balance, others: (Self, Self)) -> (Self, Self) {
		let (a, b) = self.split(amount);
		(a.merge(others.0), b.merge(others.1))
}
```
So it splits (sub) `rewards_credit` by `total_rewards % number_of_chains` then merge (add) it with `reminder_credit`. It returns the result and the remaining funds.

The merge_split using `CreditOf::<T>::zero()` is a bit confusing. I would prefer using split and merge directly.